centreon/src/Centreon/Application/Controller/Configuration/MonitoringServerController.php

```diff
@@ -216,8 +216,19 @@ private function execute(callable $callable): void
          */
         $user = $this->getUser();
         try {
-            if (! $user->isAdmin() && ! $user->hasRole(Contact::ROLE_GENERATE_CONFIGURATION)) {
-                throw new AccessDeniedException('Insufficient rights (required: ROLE_GENERATE_CONFIGURATION)');
+            if (! $user->isAdmin()) {
+                if (! $user->hasRole(Contact::ROLE_GENERATE_CONFIGURATION)) {
+                    throw new AccessDeniedException('Insufficient rights (required: ROLE_GENERATE_CONFIGURATION)');
+                }
+
+                if (
+                    ! $user->hasTopologyRole(Contact::ROLE_CONFIGURATION_MONITORING_SERVER_READ)
+                    && ! $user->hasTopologyRole(Contact::ROLE_CONFIGURATION_MONITORING_SERVER_READ_WRITE)
+                ) {
+                    throw new AccessDeniedException(
+                        'Insufficient rights (required: ROLE_CONFIGURATION_MONITORING_SERVER_READ or ROLE_CONFIGURATION_MONITORING_SERVER_READ_WRITE)'
+                    );
+                }
             }
```

Comment:
You can simplify the complexity by removing the first condition, as an administrator has all the roles.
```suggestion
            if (! $user->hasRole(Contact::ROLE_GENERATE_CONFIGURATION)) {
                throw new AccessDeniedException('Insufficient rights (required: ROLE_GENERATE_CONFIGURATION)');
            }

            if (
                ! $user->hasTopologyRole(Contact::ROLE_CONFIGURATION_MONITORING_SERVER_READ)
                && ! $user->hasTopologyRole(Contact::ROLE_CONFIGURATION_MONITORING_SERVER_READ_WRITE)
            ) {
                throw new AccessDeniedException(
                    'Insufficient rights (required: ROLE_CONFIGURATION_MONITORING_SERVER_READ or ROLE_CONFIGURATION_MONITORING_SERVER_READ_WRITE)'
                );
            }
```